onnxscript/autocast.py

```diff
@@ -1,30 +1,113 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+
 from __future__ import annotations
 
 from typing import Any, Callable, Optional
 
 import numpy as np
+import onnx
+from onnx import helper, numpy_helper
 from onnx.defs import OpSchema
 
 from onnxscript import tensor, values
 
+# Conversions from python values to ONNX are used by both the script converter as well
+# as the eager-mode runtime and both need to be consistent. The script converter converts
+# python values into ONNX TensorProto, while the runtime converts python values into
+# ONNXScript runtime's value-representation (based on Tensor).
+
+
+# Utilities to convert a python value to TensorProto (for use by the script converter)
+
 
-def cast_scalar_to_tensor(x, dtype=None):
-    """Promotes scalar values (bool, int, float) into tensors of rank zero.
-    The optional argument dtype specifies the desired np.dtype of the tensor.
-    Other types of inputs are returned as is.
+def py_type_to_onnx_type(pytype: type):
+    if pytype is bool:
+        return onnx.TensorProto.BOOL
+    if pytype is int:
+        return onnx.TensorProto.INT64
+    if pytype is float:
+        return onnx.TensorProto.FLOAT
+    if pytype is str:
+        return onnx.TensorProto.STRING
+    raise ValueError(f"Tensor element of type {pytype} not supported")
+
+
+def pyvalue_to_onnx_tensor(tensor_name: str, pyvalue):
+    if isinstance(pyvalue, np.ndarray):
+        return numpy_helper.from_array(pyvalue, tensor_name)
+    if isinstance(pyvalue, list):
+        if len(pyvalue) == 0:
+            raise ValueError("Cannot convert an empty list to tensor")
+        pytype = type(pyvalue[0])
+        if not all(isinstance(e, pytype) for e in pyvalue):
+            raise ValueError(
+                "Cannot convert an list with elements of different types to tensor"
+            )
+        return helper.make_tensor(
+            tensor_name,
+            py_type_to_onnx_type(pytype),
+            [len(pyvalue)],
+            pyvalue,
+        )
+    onnx_type = py_type_to_onnx_type(type(pyvalue))
+    if onnx_type is onnx.TensorProto.BOOL:
+        return helper.make_tensor(tensor_name, onnx_type, [], [int(pyvalue)])
+    if onnx_type is onnx.TensorProto.STRING:
+        return helper.make_tensor(tensor_name, onnx_type, [], vals=[pyvalue.encode("utf-8")])
+
+    return helper.make_tensor(tensor_name, onnx_type, [], [pyvalue])
+
+
+# Utilities to convert python values into onnxscript tensors.
+
+
+def promotable(x):
```

Comment:
```suggestion
def _promotable(x) -> bool:
```

nit. Also should it be private for now?